File: CSharp/HtmlWalker/Tag.Util/Text.cs
using System;
using HtmlWalker.Platform;
using HtmlWalker.Tag.Base;
using HtmlWalker.Visitor;
using HtmlWalker.Core;

namespace HtmlWalker.Tag.Util
{
    /// <summary>
    ///   This class holds string text and functions as a tag adapter.
    ///
    ///   While this class implements the ITag interface it houses text,
    ///   not an HTML tag.It purpose is to act as a tag adapter for
    ///   specific platforms that need it.
    ///   
    ///   Implements <see cref="ContentUtilTag"/>
    /// </summary>
    /// <author>Robert R Flores</author>
    public class Text : ContentUtilTag
    {
        /// <summary>
        ///   Instantiates a new text adapter
        /// </summary>
        /// <param name="factory">the tag that owns the text</param>
        /// <param name="text">the text string</param>
        public Text(WalkerPlatform.WalkerFactory factory, string text) : base(factory, factory.CreateText(text)) { }

        /// <summary>
        ///   Constructor for housing a existing text
        /// </summary>
        /// <param name="factory">used to produce objects as per the specific platform</param>
        /// <param name="obj">object managed by the adapter</param>
        public Text(WalkerPlatform.WalkerFactory factory, object obj) : base(factory, obj) { }

        #region ITag implementations

        /// <summary>
        ///   Implements <see cref="ITag.Visit(ITagVoidVisitor)"/>
        /// </summary>
        public override void Visit(ITagVoidVisitor visitor) { visitor.Visit(this); }

        /// <summary>
        ///   Implements <see cref="ITag.Visit{T}(ITagGenericVisitor{T})"/>
        /// </summary>
        public override T Visit<T>(ITagGenericVisitor<T> visitor) { return visitor.Visit(this); }

        /// <summary>
        ///  Override <see cref="BaseTag.Name"/> for one that identifies it utility function
        /// </summary>
        public override string Name => "#text";

        /// <summary>
        ///   Implements <see cref="ITag.TextValue"/>
        /// </summary>
        public override string TextValue
        {
            get
            {
                if (Factory.ConvertToEntities)
                    return EscHandler.Instance.Escape(Api.GetText(Tag));
                else
                    return Api.GetText(Tag);
            }

            set
            {
                if (Factory.ConvertToEntities)
                    Api.SetText(Tag, EscHandler.Instance.Escape(value));
                else
                    Api.SetText(Tag, value);
            }
        }

        /// <summary>
        ///   Implements <see cref="ITag.OwnedTags"/>
        /// </summary>
        public override TagList OwnedTags => null;

        /// <summary>
        ///   Implement <see cref="ITag.OwnedText"/>
        /// </summary>
        public override string OwnedText => TextValue;

        #endregion

        /// <summary>
        ///   Overrides <see cref="ToString"/>
        /// </summary>
        /// <returns>'text' owned by tag</returns>
        public override string ToString() => TextValue;
    }
}

File: Java/htmlwalker/src/main/java/htmlwalker/tag/util/Text.java
/*
 * 
 */
package htmlwalker.tag.util;

import htmlwalker.core.EscHandler;
import htmlwalker.exception.HtmlWalkerException;
import htmlwalker.platform.WalkerFactory;
import htmlwalker.tag.base.ContentUtilTag;
import htmlwalker.tag.base.TagList;
import htmlwalker.visitor.ITagGenericVisitor;
import htmlwalker.visitor.ITagVoidVisitor;

/**
 * This class holds string text and functions as a tag adapter.
 * 
 * While this class implements the ITag interface it houses text,
 * not an HTML tag. It purpose is to act as a tag adapter for
 * specific platforms that need it.
 * 
 * @author Robert Flores
 */
public class Text extends ContentUtilTag
{

	/**
	 * Instantiates a new text adapter.
	 *
	 * @param owner - the tag that owns the text
	 * @param text - the text string
	 */
	public Text(WalkerFactory factory, String text) { super(factory, factory.createText(text)); }
	
	public Text(WalkerFactory factory, Object tag) { super(factory, tag); }
	
	/**
	 * @see HtmlWalker.ITag#visit(HtmlWalker.ITagVoidVisitor)
	 */
	@Override
	public void visit(ITagVoidVisitor visitor) throws HtmlWalkerException { visitor.visit(this); }

	/**
	 * @see HtmlWalker.ITag#visit(HtmlWalker.ITagGenericVisitor)
	 */
	@Override
	public <T> T visit(ITagGenericVisitor<T> visitor) throws HtmlWalkerException { return visitor.visit(this); }

	/**
	 * @see HtmlWalker.ITag#name()
	 */
	@Override
	public String name() { return "#text"; }

	@Override
	public String getTextValue()
	{
        if (this.factory.convertToEntities)
            return EscHandler.theInstance.escape(this.api.getText(tag()));
        else
            return this.api.getText(tag());
	}

	@Override
	public void setTextValue(String text)
	{
		if (this.factory.convertToEntities)
			this.api.setText(tag(), EscHandler.theInstance.escape(text));
		else
			this.api.setText(tag(), text);
	}

	@Override
	public TagList ownedTags() { throw new UnsupportedOperationException(); }
	
	/**
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString() { return getTextValue(); }

	@Override
	public String ownedText() { return getTextValue(); }

}
